### PR TITLE
[flang][openacc] Fixed bounds for full array sections.

### PR DESCRIPTION
**Pull request by vzakhari**
We have to use 0-based lowerbound/upperbound for data operands like `a(:)`.

### CODE REVIEW

**Comment by vzakhari** [comment]
As I understand, this should affect OpenMP as well, but none of OpenMP LIT tests failed.

File: flang/test/Lower/OpenACC/acc-enter-data.f90 [hunk 1]
```diff
@@ -288,10 +288,11 @@ subroutine acc_enter_data_dummy(a, b, n, m)
 !CHECK: acc.enter_data dataOperands(%[[CREATE1]] : !fir.ref<!fir.array<?xf32>>)
 
   !$acc enter data create(b(:))
+!CHECK: %[[ZERO:.*]] = arith.constant 0 : index
 !CHECK: %[[ONE:.*]] = arith.constant 1 : index
 !CHECK: %[[LBEXT:.*]] = arith.addi %[[EXT_B]], %[[N_IDX]] : index
 !CHECK: %[[UB:.*]] = arith.subi %[[LBEXT]], %[[ONE]] : index
-!CHECK: %[[BOUND1:.*]] = acc.bounds lowerbound(%[[N_IDX]] : index) upperbound(%[[UB]] : index) extent(%[[EXT_B]] : index) stride(%[[ONE]] : index) startIdx(%[[N_IDX]] : index)
+!CHECK: %[[BOUND1:.*]] = acc.bounds lowerbound(%[[ZERO]] : index) upperbound(%[[UB]] : index) extent(%[[EXT_B]] : index) stride(%[[ONE]] : index) startIdx(%[[N_IDX]] : index)
```
Comment (clementval) [inline flang/test/Lower/OpenACC/acc-enter-data.f90:295]: If we set the lb to zero then the ub needs to be adjusted as well otherwise it doesn't match.

**Review by clementval — APPROVED** [review]
LGTM. Thanks!

**Comment by agozillon** [comment]
> As I understand, this should affect OpenMP as well, but none of OpenMP LIT tests failed.

It does, but I think OpenACC tests the bounds generation a little more strictly than the tests I've added so far for the bounds, which are primarily just some lit tests to make sure fortran generated the bounds as expected in certain cases and then using those bounds to test further lowering to LLVM-IR. 

However, if something was going to fail it would likely fail an openmp/libomptarget/offloading test e.g. https://github.com/llvm/llvm-project/blob/main/openmp/libomptarget/test/offloading/fortran/basic-target-region-3D-array-section.f90 but they're not really testable without an AMD GPU and I'm fairly certain none of the upstream buildbots run these tests at the moment so it won't fail a buildbot (nothing that compiles Flang and runs OpenMP tests with AMD GPU offload). 

I could test this PR tomorrow and if it breaks anything create another PR that can be approved and merged at the same time by you or incorporated into this patch. Or if we're in a bit of a hurry to land this, I can test they all work reasonably after this patch lands and create a fix if they do not. Whatever works best.

**Comment by vzakhari** [comment]
Thank you for the reply, @agozillon! I will appreciated it if you could test the patch end-to-end!  I can postpone the merge for a day, no problem.

**Review by agozillon — APPROVED** [review]
Works as expected, as does testing with another test case using the syntax a(:) that we don't currently cover! Thank you very much for your patience and the great fix. It all LGTM. If anything unexpected crops up that is OpenMP related just ping me and I am happy to look into it.

**Pull request merged**